drivers/hmis/app/models/hmis/user.rb

```diff
@@ -54,6 +57,23 @@ def load_effective_permissions
       send(permission)
     end
 
+    define_method("#{permission}_for?") do |entity|
+      joins_association = nil
+      joins_association = :projects if entity.is_a?(Hmis::Hud::Project)
+      joins_association = :organizations if entity.is_a?(Hmis::Hud::Organization)
+
+      raise "Invalid entity '#{entity.class.name}'" unless joins_association.present?
+      return false unless send("#{permission}?")
+
+      access_groups.
+        merge(
+          Hmis::AccessGroup.joins(:access_controls).
+            where(id: Hmis::GroupViewableEntity.where(entity_type: entity.class.name, entity_id: entity.id).pluck(:access_group_id)).
+            merge(Hmis::AccessControl.where(role_id: roles.where(permission => true).pluck(:id))),
+        )
+        .exists?
+    end
```

Comment:
I _think_ this logic makes sense to me, but I'm finding it confusing to read.

Would the below logic achieve the same result?

```
     # All access groups that include the specified entity
     access_group_ids = Hmis::GroupViewableEntity.where(entity_type: entity.class.name, entity_id: entity.id).pluck(:access_group_id)
    
     # All roles that include the specified permission
     role_ids = Hmis::Role.where(permission => true)

     access_controls.where(access_group_id: access_group_ids, role_id: role_ids).exists?
```

Reply:
Actually, there is an issue here. `Hmis::GroupViewableEntity.where(entity_type: entity.class.name, entity_id: entity.id)` isn't the right way to find _any GroupViewableEntity that grants access to this <project|organization>_. It will only return GroupViewableEntities that are explicitly assigned to the specified type. It incorrectly excludes group viewable entities that grant access to a project by way of an organization, data source, or project group.


In other words: you should be able to call `user.permission_for?(organization)` and `user.permission_for?(project)`  and have it return `true` if the user is part of an ACL that belongs to a _group_ that grants access to the _data source_ that the specified project/organization belong to.

Maybe a clean way to do this would be to add scopes for `:includes_project` and `:includes_organization` so we can say _"give me all the GroupViewableEntities that grant access to this record."_ Then we could do something like:

```
# All access groups that include the specified project
access_group_ids = Hmis::GroupViewableEntity.includes_project(entity).pluck(:access_group_id) if entity.is_a?(Hmis::Hud::Project)
access_group_ids = Hmis::GroupViewableEntity.includes_organization(entity).pluck(:access_group_id) if entity.is_a?(Hmis::Hud::Organization)
```

Reply:
We may want to chat with @awisema to get some feedback on this. I don't know if we do anything similar with GroupViewableEntities in the warehouse. The closest thing I can think of is the Project `viewable_by` scope which evaluates all entity levels to find all viewable projects.